### PR TITLE
20_2 TreeList Column Chooser: rewrite

### PR DESCRIPTION
**Pull request by pavel-2020**
[Trello](https://trello.com/c/Bi9lS4Ee/1654-treelist-demos-column-customization-4-demos)

### CODE REVIEW

File: JSDemos/Demos/TreeList/ColumnChooser/description.md [hunk 1]
```diff
@@ -1 +1,13 @@
-The Column Chooser is&nbsp;an&nbsp;element that allows a&nbsp;user to&nbsp;alter the set of&nbsp;visible columns at&nbsp;runtime. To&nbsp;open the column chooser, click the _&laquo;Column Chooser&raquo;_ button at&nbsp;the top of&nbsp;TreeList. Now, you can drag and drop column headers to&nbsp;and from the TreeList, thus showing or&nbsp;hiding columns. Use the column chooser&rsquo;s search box to&nbsp;quickly find the required column.
\ No newline at end of file
+The column chooser allows users to hide and unhide columns at runtime. To enable it, assign **true** to the **columnChooser**.[enabled](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#enabled) property.
```
Comment (AlexJameson) [inline JSDemos/Demos/TreeList/ColumnChooser/description.md]: >unhide

I'm not sure that this word exists. I had an email conversation with teamdocs long ago about a similar phrase, and the suggestion was the following:

```suggestion
The column chooser allows users to hide or display columns at runtime. To enable it, assign **true** to the **columnChooser**.[enabled](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#enabled) property. 
```

File: JSDemos/Demos/TreeList/ColumnChooser/description.md [hunk 2]
```diff
@@ -1 +1,13 @@
-The Column Chooser is&nbsp;an&nbsp;element that allows a&nbsp;user to&nbsp;alter the set of&nbsp;visible columns at&nbsp;runtime. To&nbsp;open the column chooser, click the _&laquo;Column Chooser&raquo;_ button at&nbsp;the top of&nbsp;TreeList. Now, you can drag and drop column headers to&nbsp;and from the TreeList, thus showing or&nbsp;hiding columns. Use the column chooser&rsquo;s search box to&nbsp;quickly find the required column.
\ No newline at end of file
+The column chooser allows users to hide and unhide columns at runtime. To enable it, assign **true** to the **columnChooser**.[enabled](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#enabled) property. 
+
+To open the column chooser, click the button in the toolbar above the TreeList component. The opened popup panel lists the headers of hidden columns. In this demo, you can specify the way users interact with this panel to hide/unhide columns. For this, use the drop-down list below the TreeList component to set the [mode](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#mode) property to one of the following values:
```
Comment (AlexJameson) [inline JSDemos/Demos/TreeList/ColumnChooser/description.md]: 1. >click the button in the toolbar above the TreeList component.

It is a user who clicks the button, isn't it? Also, the word 'component' can be removed.

2. >The opened popup panel lists the headers of hidden columns.

This works only in the 'Drag and Drop' mode, right? Can we also remove the word 'panel'?

3. >to hide/unhide columns

The same correction as above

File: JSDemos/Demos/TreeList/ColumnChooser/description.md [hunk 3]
```diff
@@ -1 +1,13 @@
...
+
+* **"dragAndDrop"**
+Users have to drag and drop column headers to and from the column chooser panel.
+
+* **"select"**
+Users have to choose checkboxes to select visible columns.
```
Comment (AlexJameson) [inline JSDemos/Demos/TreeList/ColumnChooser/description.md]: Seem like we can remove 'have to' and nothing bad happens. Also, the word 'choose' is misused here:

```suggestion
* **"dragAndDrop"**
Users drag and drop column headers to and from the column chooser panel.

* **"select"**
Users mark checkboxes to select visible columns.
```

File: JSDemos/Demos/TreeList/ColumnChooser/description.md [hunk 4]
```diff
@@ -1 +1,13 @@
...
+
+If the [allowSearch](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#allowSearch) is set to **true**, you can search through the column chooser values. In this demo, use the checkbox below the TreeList to toggle this feature.
```
Comment (AlexJameson) [inline JSDemos/Demos/TreeList/ColumnChooser/description.md]:
```suggestion
If the [allowSearch](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#allowSearch) property is set to **true**, users can search through the column chooser values. In this demo, use the checkbox below the TreeList to toggle this feature.
```

File: JSDemos/Demos/TreeList/ColumnChooser/description.md [hunk 5]
```diff
@@ -1 +1,13 @@
...
+
+You can also hide certain columns programmatically. For this, set **columns[]**.[visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visible) to **false**. If a column should never be hidden, assign **false** to **columns[]**.[allowHiding](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#allowHiding).
```
Comment (AlexJameson) [inline JSDemos/Demos/TreeList/ColumnChooser/description.md]: The word 'certain' is banned by Ray. I can't find the exact place, so [here](https://teams.microsoft.com/l/message/19:78e9fcaa3c734fc3a95cdfe1bbb57f41@thread.skype/1586327699676?tenantId=e4d60396-9352-4ae8-b84c-e69244584fa4&groupId=5afdc900-1758-4beb-8368-d4d641dc99ea&parentMessageId=1586327699676&teamName=Let%20us%20speak%20English%20and%20discuss%20it%20here!&channelName=General&createdTime=1586327699676) is the first post with replacement I found

```suggestion
You can also hide specific columns programmatically. For this, set **columns[]**.[visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visible) to **false**. If a column should never be hidden, assign **false** to **columns[]**.[allowHiding](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#allowHiding). 
```

File: JSDemos/Demos/TreeList/ColumnChooser/description.md [hunk 6]
```diff
@@ -1 +1,13 @@
...
+
+To open the column chooser, users should click the button in the toolbar above the TreeList. The way users show and hide columns depends on the **columnChooser**.[mode](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#mode):
+
+* *"dragAndDrop"*              
+Users drag and drop column headers to and from the column chooser.
+
+* *"select"*         
+Users select and unselect checkboxes with column names.
+
+If the [allowSearch](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#allowSearch) property is set to **true**, users can search through the column chooser values. In this demo, use the checkbox below the TreeList to toggle this feature.
```
Comment (RomanTsukanov) [inline JSDemos/Demos/TreeList/ColumnChooser/description.md]:
```suggestion
If the [allowSearch](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#allowSearch) property is set to **true**, users can search through column names. In this demo, use the checkbox below the TreeList to toggle this feature.
```

File: JSDemos/Demos/TreeList/ColumnChooser/description.md [hunk 7]
```diff
@@ -1 +1,13 @@
...
+
+You can also hide specific columns programmatically. For this, set **columns[]**.[visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visible) to **false**. If a column should never be hidden, assign **false** to **columns[]**.[allowHiding](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#allowHiding).
```
Comment (RomanTsukanov) [inline JSDemos/Demos/TreeList/ColumnChooser/description.md]:
```suggestion
You can also hide specific columns programmatically. For this, set the **columns[]**.[visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visible) property to **false**. If a column should never be hidden, assign **false** to the **columns[]**.[allowHiding](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#allowHiding) property. 
```

**Review by RomanTsukanov — CHANGES_REQUESTED** [review]

**Review by RomanTsukanov — APPROVED** [review]

File: JSDemos/Demos/TreeList/ColumnChooser/description.md [hunk 8]
```diff
@@ -1 +1,13 @@
...
\ No newline at end of file
+The column chooser allows users to hide and display columns at runtime. To enable it, assign **true** to the **columnChooser**.[enabled](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#enabled) property. 
+
+To open the column chooser, users should click the button in the toolbar above the TreeList. The way users show and hide columns depends on the **columnChooser**.[mode](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#mode):
+
+* *"dragAndDrop"*              
+Users drag and drop column headers to and from the column chooser.
+
+* *"select"*         
+Users select and unselect checkboxes with column names.
```
Comment (Abadzhev) [inline JSDemos/Demos/TreeList/ColumnChooser/description.md]: Microsoft recommends "deselect" for checkboxes. It's weird, I know... 

```suggestion
Users select and deselect checkboxes with column names.
```

File: JSDemos/Demos/TreeList/ColumnChooser/description.md [hunk 9]
```diff
@@ -1 +1,13 @@
...
+
+If the [allowSearch](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#allowSearch) property is set to **true**, users can search through column names. In this demo, use the checkbox below the TreeList to toggle this feature.
+
+You can also hide specific columns programmatically. For this, set the **columns[]**.[visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visible) property to **false**. If a column should never be hidden, assign **false** to the **columns[]**.[allowHiding](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#allowHiding) property.
```
Comment (Abadzhev) [inline JSDemos/Demos/TreeList/ColumnChooser/description.md]: I moved part of this to the top - looks to me that it makes more sense in the same paragraph with Column Chooser description. Looks to me like "how to change in code" and "how to not allow users to hide a column" are different thoughts - shouldn't really be in the same paragraph. 

```suggestion
To hide a column in code, set the **columns[]**.[visible](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#visible) property to **false**. 
```

File: JSDemos/Demos/TreeList/ColumnChooser/description.md [hunk 10]
```diff
@@ -1 +1,13 @@
-The Column Chooser is&nbsp;an&nbsp;element that allows a&nbsp;user to&nbsp;alter the set of&nbsp;visible columns at&nbsp;runtime. To&nbsp;open the column chooser, click the _&laquo;Column Chooser&raquo;_ button at&nbsp;the top of&nbsp;TreeList. Now, you can drag and drop column headers to&nbsp;and from the TreeList, thus showing or&nbsp;hiding columns. Use the column chooser&rsquo;s search box to&nbsp;quickly find the required column.
\ No newline at end of file
+The column chooser allows users to hide and display columns at runtime. To enable it, assign **true** to the **columnChooser**.[enabled](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#enabled) property.
```
Comment (Abadzhev) [inline JSDemos/Demos/TreeList/ColumnChooser/description.md]: "Hide and display" - in this phrase "display" doesn't look like an opposite to hide. "Hide" implies a change of state from visible to invisible. "Display" doesn't really mean the columns were previously hidden. If a control displays something, it doesn't mean this something was hidden before. So I'd maybe use "change visibility" or "hide and make visible". Even "hide and show" is better, I guess. 

And then I think the line about columns that can always remain visible works better here (I moved it up from the very end.)

```suggestion
If you want users to change column visibility at runtime, enable the Column Chooser. Set the **columnChooser**.[enabled](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#enabled) property to **true**.  Note that you can make sure that certain columns always stay visible. To do so, set the **columns[]**.[allowHiding](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#allowHiding) property to **false**. 
```

File: JSDemos/Demos/TreeList/ColumnChooser/description.md [hunk 11]
```diff
@@ -1 +1,13 @@
-The Column Chooser is&nbsp;an&nbsp;element that allows a&nbsp;user to&nbsp;alter the set of&nbsp;visible columns at&nbsp;runtime. To&nbsp;open the column chooser, click the _&laquo;Column Chooser&raquo;_ button at&nbsp;the top of&nbsp;TreeList. Now, you can drag and drop column headers to&nbsp;and from the TreeList, thus showing or&nbsp;hiding columns. Use the column chooser&rsquo;s search box to&nbsp;quickly find the required column.
\ No newline at end of file
+If you want users to change column visibility at runtime, enable the Column Chooser. Set the **columnChooser**.[enabled](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#enabled) property to **true**.  Note that you can make sure that certain columns always stay visible. To do so, set the **columns[]**.[allowHiding](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columns/#allowHiding) property to **false**. 
+
+To open the column chooser, users should click the button in the toolbar above the TreeList. The way users show and hide columns depends on the **columnChooser**.[mode](/Documentation/ApiReference/UI_Components/dxTreeList/Configuration/columnChooser/#mode):
```
Comment (albertov05) [inline JSDemos/Demos/TreeList/ColumnChooser/description.md:3]: Should "`the **columnChooser**.[mode]`" be "`the **columnChooser** [mode]`" (with a space instead of a period)?

Or, if `**columnChooser**.[mode]` is a property name, do you mean "`the **columnChooser**.[mode] property value`"?

Comment (RomanTsukanov) [reply JSDemos/Demos/TreeList/ColumnChooser/description.md:3]: No, the dot designates that the property `mode` is inside the `columnChooser` object.

Comment (albertov05) [reply JSDemos/Demos/TreeList/ColumnChooser/description.md:3]: Okay, thank you.

**Pull request merged**